src/core/execute.c

```diff
@@ -5831,36 +5861,68 @@ int exec_spawn(Unit *unit,
+         * and ensure all memory is shared. The child immediately execs the new binary so the delay should
+         * be minimal. Once glibc provides a clone3 wrapper we can switch to that, and clone directly in the
+         * target cgroup. */
 
-        if (pid == 0) {
-                int exit_status = EXIT_SUCCESS;
+        serialization_fd = open_serialization_fd("sd-executor-state");
+        if (serialization_fd < 0)
+                return log_unit_error_errno(unit, serialization_fd, "Failed to open serialization fd: %m");
 
-                r = exec_child(unit,
-                               command,
-                               context,
-                               params,
-                               runtime,
-                               cgroup_context,
-                               &exit_status);
+        f = take_fdopen(&serialization_fd, "w+");
+        if (!f)
+                return log_unit_error_errno(unit, serialization_fd, "Failed to open serialization stream: %m");
 
-                if (r < 0) {
-                        const char *status =
-                                exit_status_to_string(exit_status,
-                                                      EXIT_STATUS_LIBC | EXIT_STATUS_SYSTEMD);
+        fdset = fdset_new();
+        if (!fdset)
+                return log_oom();
 
-                        log_unit_struct_errno(unit, LOG_ERR, r,
-                                              "MESSAGE_ID=" SD_MESSAGE_SPAWN_FAILED_STR,
-                                              LOG_UNIT_INVOCATION_ID(unit),
-                                              LOG_UNIT_MESSAGE(unit, "Failed at step %s spawning %s: %m",
-                                                               status, command->path),
-                                              "EXECUTABLE=%s", command->path);
-                }
+        r = unit_serialize(unit, f, fdset, /* switching_root */ false);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize unit '%s': %m", unit->id);
 
-                _exit(exit_status);
-        }
+        r = exec_command_serialize(command, f);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecCommand: %m");
+
+        r = exec_parameters_serialize(params, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecParameters: %m");
+
+        r = exec_runtime_serialize(runtime, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecRuntime: %m");
+
+        r = exec_context_serialize(context, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecContext: %m");
+
+        r = exec_cgroup_context_serialize(cgroup_context, f);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize CGroupContext: %m");
```

Comment:
maybe split these series of serialization calls out into a helper function of its own, too. it has this repetitive feel even though it is of course not repetitive